nixos/top-level: change extraDependencies from packages to paths

ncfavier opened the pull request:
I was [once again](https://github.com/NixOS/nixpkgs/pull/163443) trying to work around https://github.com/NixOS/nix/issues/3995 by (recursively) adding my flake inputs to `system.extraDependencies`; this time `nixd`, which depends on `flake-parts`, which depends on the `lib` subdirectory of `nixpkgs`.

This fails because the subdirectory flake does not pass `types.package.check`: it's not a derivation, and it's not a store path either (it's a subdirectory of a store path).

Change the type to `path`, because there's no "store path or descendant" type. Honestly I'm not sure why anyone would ever use `types.package` at this point.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes (or backporting 23.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

ncfavier committed:
nixos/top-level: change extraDependencies from package to pathInStore

Allows adding subdirectory flake inputs that aren't top-level store paths.

roberth commented:
> because there's no "store path or descendant" type.

We could call that `types.pathInStore` perhaps.


> Honestly I'm not sure why anyone would ever use `types.package` at this point.

Packages and paths *should* be treated differently, because packages can have multiple outputs. I'm pretty sure `extraDependencies` doesn't treat packages like packages. What even is the implied usage for those packages? Should it keep `.doc`? Does that even make sense? We should probably rename it to `extraPaths`, so that you can guess the behavior correctly, as it effectively just takes the `outPath`.

roberth commented on nixos/modules/system/activation/top-level.nix:
```diff
@@ -260,10 +260,10 @@ in
 
     system.extraDependencies = mkOption {
-      type = types.listOf types.package;
+      type = types.listOf types.path;
```
Wouldn't this allow values like `"/var"`? I think we should keep making sure they're in the store at least.

ncfavier committed:
lib/types: add pathInStore

ncfavier commented:
Added `pathInStore`.

I'm not sure `extraPaths` is less ambiguous than `extraDependencies` so I've left the option name alone.

ncfavier commented on lib/types.nix:
```diff
@@ -491,6 +492,14 @@ rec {
+      name = "pathInStore";
+      description = "path in the Nix store";
+      descriptionClass = "noun";
+      check = x: isStringLike x && lib.hasPrefix builtins.storeDir (toString x);
```
(I wish we had a proper subdirectory test in the path library.)

ncfavier replied:
Actually we probably only want to accept *strict* subdirectories here, so I'll just add a slash.

roberth committed:
lib/tests/modules.sh: Test types.pathInStore

Add missing test cases. I think the .links case should be rejected
even though it's technically a path in the store.

roberth commented on lib/tests/modules.sh:
```diff
@@ -61,6 +61,15 @@ checkConfigError() {
+checkConfigError 'A definition for option .* is not of type .path in the Nix store.. Definition values:\n\s*- In .*: ""' config.pathInStore.bad1 ./types.nix
+checkConfigError 'A definition for option .* is not of type .path in the Nix store.. Definition values:\n\s*- In .*: "/nix/store"' config.pathInStore.bad2 ./types.nix
+checkConfigError 'A definition for option .* is not of type .path in the Nix store.. Definition values:\n\s*- In .*: "/nix/store/"' config.pathInStore.bad3 ./types.nix
+checkConfigError 'A definition for option .* is not of type .path in the Nix store.. Definition values:\n\s*- In .*: "/nix/store/.links"' config.pathInStore.bad4 ./types.nix
```
Technically "valid" but this would never be a valid store path, which I think we should consider as a condition for being in the store, from the Nix language perspective.

ncfavier committed:
lib/tests: remove experimental-features

Now that the lib is tested with Nix 2.3, this isn't needed any more and
causes warnings.
